src/swarm-lang/Swarm/Language/LSP/Hover.hs

```diff
@@ -88,60 +88,86 @@ posToRange myRope foundSloc = do
-  SLam lv _ s -> d (locVarToSyntax' lv $ getInnerType ty) <|> d s
-  SApp s1 s2 -> d s1 <|> d s2
-  SLet _ _ lv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> d (locVarToSyntax' lv lty) <|> d s1 <|> d s2
-  SBind mlv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> (mlv >>= d . flip locVarToSyntax' (getInnerType lty)) <|> d s1 <|> d s2
-  STydef typ typBody _ti s1 -> d s1 <|> Just (locVarToSyntax' (tdVarName <$> typ) $ fromPoly typBody)
-  SPair s1 s2 -> d s1 <|> d s2
-  SDelay s -> d s
-  SRcd m -> asum . map d . catMaybes . M.elems $ m
-  SProj s1 _ -> d s1
-  SAnnotate s _ -> d s
-  SRequirements _ s -> d s
-  SParens s -> d s
-  -- atoms - return their position and end recursion
-  TUnit -> Nothing
-  TConst {} -> Nothing
-  TDir {} -> Nothing
-  TInt {} -> Nothing
-  TText {} -> Nothing
-  TBool {} -> Nothing
-  TVar {} -> Nothing
-  TStock {} -> Nothing
-  TRequire {} -> Nothing
-  TType {} -> Nothing
-  -- these should not show up in surface language
-  TRef {} -> Nothing
-  TRobot {} -> Nothing
-  TAntiInt {} -> Nothing
-  TAntiText {} -> Nothing
-  TAntiSyn {} -> Nothing
-  SSuspend {} -> Nothing
+  NonEmpty (Syntax' ty)
+pathToPosition s0 pos = s0 :| innerPath s0
  where
+  innerPath :: Syntax' ty -> [Syntax' ty]
+  innerPath (Syntax' _ t _ ty) = case t of
+    SLam lv _ s -> d (locVarToSyntax' lv $ getInnerType ty) <|> d s
+    SApp s1 s2 -> d s1 <|> d s2
+    SLet _ _ lv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> d (locVarToSyntax' lv lty) <|> d s1 <|> d s2
+    SBind mlv _ _ _ s1 s2 -> bindPath mlv s1 s2
+    STydef typ typBody _ti s1 -> d s1 <|> [locVarToSyntax' (tdVarName <$> typ) $ fromPoly typBody]
+    SPair s1 s2 -> d s1 <|> d s2
+    SDelay s -> d s
+    SRcd m -> asum . map d . catMaybes . M.elems $ m
+    SProj s1 _ -> d s1
+    SAnnotate s _ -> d s
+    SRequirements _ s -> d s
+    SParens s -> d s
+    -- atoms - return their position and end recursion
+    TUnit -> mempty
+    TConst {} -> mempty
+    TDir {} -> mempty
+    TInt {} -> mempty
+    TText {} -> mempty
+    TBool {} -> mempty
+    TVar {} -> mempty
+    TStock {} -> mempty
+    TRequire {} -> mempty
+    TType {} -> mempty
+    -- these should not show up in surface language
+    TRef {} -> mempty
+    TRobot {} -> mempty
+    TAntiInt {} -> mempty
+    TAntiText {} -> mempty
+    TAntiSyn {} -> mempty
+    SSuspend {} -> mempty
+
+  bindPath :: Maybe LocVar -> Syntax' ty -> Syntax' ty -> [Syntax' ty]
+  bindPath mlv s1@(Syntax' _ _ _ lty) s2 = mlvToSyntax mlv <|> d s1 <|> d s2
+   where
+    mlvToSyntax :: Maybe LocVar -> [Syntax' ty]
+    mlvToSyntax = \case
+      Just lv -> d $ locVarToSyntax' lv (getInnerType lty)
+      Nothing -> []
+
   d = descend pos
 
+  -- try and decend into the syntax element if it is contained with position
+  descend ::
+    ExplainableType ty =>
+    -- \| position
+    Int ->
+    -- \| next element to inspect
+    Syntax' ty ->
+    [Syntax' ty]
+  descend p s1@(Syntax' l1 _ _ _) = do
+    guard $ withinBound p l1
+    return $ narrowToPosition s1 p
```

Comment:
I think `descend` must now call `innerPath` rather than `narrowToPosition`, and should have a type like `Syntax' ty -> Maybe [Syntax' ty]`, just like `innerPath`.  This is probably the place to add the root node onto the beginning of the path, so something like
```
  guard $ withinBound p l1
  return $ (s1 :) <$> innerPath s1 p
```
so that the `s1` node gets consed onto the front of the path returned by `pathToPosition`.  That way, as `innerPath` keeps recursively calling `descend` on its way down the tree, it will build up a path of all the syntax nodes encountered along the way.